activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -15,47 +15,37 @@ def source_reflection
         def associated_records_by_owner(preloader)
           through_scope = through_scope()
 
-          preloader.preload(owners,
+          through_preloaders = preloader.preload(owners,
                             through_reflection.name,
-                            through_scope)
+                            through_scope,
+                            should_skip_setting_target?(owners, through_reflection.name, through_scope))
 
-          through_records = owners.map do |owner|
-            center = owner.association(through_reflection.name).target
-            [owner, Array(center)]
-          end
-
-          reset_association(owners, through_reflection.name, through_scope)
```

Comment:
Just being curious is there a particular reason that you did all this rather than just calling reset_association twice. One before preloader.preload at line 18 and one here as it is. If doing so still cover all the scenario than that will reduce no. of changes in code which will simplify reviewing process.